Add additional `unsigned_to_signed` doc (#4076)
SpikeInterface/spikeinterface: doc/how_to/unsigned_to_signed.rst

Unsigned to Signed Data types
=============================

As of version 0.103.0 SpikeInterface has changed one of its defaults for interacting with

> :code:`Recording`` objects. We no longer autocast unsigned dtypes to signed implicitly. This

Reviewer: ```suggestion
:code:`Recording` objects. We no longer autocast unsigned dtypes to signed implicitly. This
```

:code: + double tick gives a rendered tick

means that some users of SpikeInterface will need to add one additional line of code to their scripts
to explicitly handle this conversion.

Why this matters?
-----------------

For those that want a deeper understanding of dtypes `NumPy provides a great explanation <https://numpy.org/doc/stable/reference/arrays.dtypes.html>`_.
For our purposes it is important to know that many pieces of recording equipment opt to store their electrophysiological data as unsigned integers,

> which provides the benefit of reduces the necessary file size. In order to convert to real units these file formats only need to store a :code:`gain`

Reviewer: ```suggestion
which provides the benefit of reducing the necessary file size. In order to convert to real units these file formats only need to store a :code:`gain`
```

solution required is to convernt unsigned integers into signed integers. Previously we did this under the hood, automatically for users that had
a :code:`Recording` object with an unsigned dtype.

We decided, however, that implicitly performing this action was not the best course of action. So from version 0.103.0, users will now explicitly
have to perform this transformation of their data. This will help users better understand how they are processing their data during an analysis
pipeline as well as better understand the provenance of their pipeline.

Using :code:`unsigned_to_signed`
--------------------------------

For users that receive an error because their :code:`Recording` is unsigned, their is one additional step that must be done:

.. code:: python

    import spikeinterface.extractors as se
    import spikeinterface.preprocessing as spre

    # Intan is an example of unsigned data
    recording = se.read_intan('path/to/my/file.rhd', stream_id='0')
    # to get a signed version of our Recording we use the following function
    recording_signed = spre.unsigned_to_signed(recording)

>     # we can use all functions that we used previously in our scripts

Reviewer: ```suggestion
    # we can now apply any functions that we used previously in our scripts, e.g.
```

    recording_filtered = spre.bandpass_filter(recording_signed)

Now with the signed dtype of the :code:`Recording` one can use a SpikeInterface pipeline as usual.

If you are curious if your :code:`Recording` is unsigned you can simply check the repr or use :code:`get_dtype()`

.. code:: python

    # the repr automatically displays the dtype
    print(recording)
    # use method on the Recording object
    print(recording.get_dtype())

> 

Reviewer: Would be good to mention what you'd expect here. E.g. "If this returns a dtype starting with a u, such as `uint16`, the recording if using unsigned int. While if this returns ... "

Author: Do you think I should just write up a parsed-literal block or you think it is sufficient to just say `uint16`?

Reviewer: Up to you. I think sufficient just to say: the possible things it can return are pretty simple.

Additional Notes
----------------

> 1) Some some sorters make use of SpikeInterface preprocessing either

Reviewer: Also "Some some" -> "some" typo!

> 1) Some some sorters make use of SpikeInterface preprocessing either

Author: Will fix the some some (haha). I didn't explicitly intend it as a list. Just numbered, but rst does tend to render non-lists relatively ugly.

Unsigned to Signed Data types
=============================

As of version 0.103.0 SpikeInterface has changed one of its defaults for interacting with
:code:`Recording` objects. We no longer autocast unsigned dtypes to signed implicitly. This

> means that some users of SpikeInterface will need to add one additional line of code to their scripts

Author: Yes it is in the example. I know Intan does this. But that's the only one I know for sure off the top of my head.  (I could do a `e.g. Intan`) but I also didn't want to preference one system.

> means that some users of SpikeInterface will need to add one additional line of code to their scripts

Reviewer 2: I think it is fine to add only one to have at least is one example. We add more if we discover them to be more. I guess we can just do a search for uint16 in neo but for this PR at least let's have an example.

> means that some users of SpikeInterface will need to add one additional line of code to their scripts

Reviewer 2: Maxwell:

https://github.com/h-mayorquin/python-neo/blob/2240a29c4366fb3ed5dacadb54837004651ee35e/neo/rawio/maxwellrawio.py#L184-L192

> means that some users of SpikeInterface will need to add one additional line of code to their scripts

Reviewer 2: I can add the list in another PR if you want @zm711. You have already done a lot with this : )

to explicitly handle this conversion.

Why this matters?
-----------------

For those that want a deeper understanding of dtypes `NumPy provides a great explanation <https://numpy.org/doc/stable/reference/arrays.dtypes.html>`_.

> For our purposes it is important to know that many pieces of recording equipment opt to store their electrophysiological data as unsigned integers,

Reviewer 2: "Many pieces" maybe being specific here would help if you know the formats that do this.

Author: "acquisition systems"? is that better? I don't necessarily want to go into depth here unless we do end up doing an exhaustive list. So I guess answering your statement above will determine what we do here.

Reviewer 2: Yes, same as above. Let's give some examples.

> which provides the benefit of reducing the necessary file size. In order to convert to real units these file formats only need to store a :code:`gain`

Reviewer 3: I don't think this is true, since `unit16` and `int16` has the same size (16 bit). The difference is that unsigned values have a twice as large range ov values, because for signed ints the first bit is the sign.

> and an :code:`offset`. Our :code:`RecordingExtractor`'s maintain the dtype that the file format utilizes, which means that some of our

Author: I don't care either way to happy to delete if you think that would be better.

:code:`RecordingExtractor`'s will have unsigned dtypes.

The problem with using unsigned dtypes is that many types of funtions (including the ones we use from SciPy) perform poorly with unsigned integers.
This is made worse by the fact that these failures are silent (i.e. no error is triggered but the operation leads to nonsensical data). So the
solution required is to convernt unsigned integers into signed integers. Previously we did this under the hood, automatically for users that had
a :code:`Recording` object with an unsigned dtype.

> We decided, however, that implicitly performing this action was not the best course of action. So from version 0.103.0, users will now explicitly

Reviewer 2: If we remember it here we can give the reasoning. Do we remember?

Author: Okay, maybe I can text Alessio tomorrow morning for a chat to see if he wants me to add. I could imagine explicit better, + better control of bit_depth (but I don't remember how the old function worked).

Reviewer 2: Yes, it is just that it is even bad (even for us) if we say we decided and we don't add a rationale. Beats the purpose of documenting this. I rather remove the first sentence instead of giving a statement of what we did if we don't give a rationale.
